### PR TITLE
add FeedInventory controller

### PR DESCRIPTION
**Pull request by dixudx**
#### What type of PR is this?
kind/feature
kind/api
kind/change

#### Which issue(s) this PR fixes:

Fixes #

### CODE REVIEW

File: pkg/controllers/apps/feedinventory/feedinventory.go [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+			if err2 != nil {
+				errCh <- err2
+				return
+			}
+			if manifests == nil {
+				errCh <- apierrors.NewNotFound(schema.GroupResource{Resource: feed.Kind}, feed.Name)
+				return
+			}
+
+			gvk, err2 := getGroupVersionKind(manifests[0].Template.Raw)
```
Comment (Garrybest) [inline pkg/controllers/apps/feedinventory/feedinventory.go:302]: Does the manifest always have only one element?

File: pkg/controllers/apps/feedinventory/feedinventory.go [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+		subLister:         subsInformer.Lister(),
+		subSynced:         subsInformer.Informer().HasSynced,
+		manifestLister:    manifestInformer.Lister(),
+		manifestSynced:    manifestInformer.Informer().HasSynced,
+		recorder:          recorder,
+		registry:          registry,
+		reservedNamespace: reservedNamespace,
+	}
+
+	// For FeedInventory, we only need to care about the changes of Subscription
```
Comment (Garrybest) [inline pkg/controllers/apps/feedinventory/feedinventory.go:90]: What if a workload has been scaled up but the subscription has not been modified? It seems that this controller does not care about the events of original workload. E.g., a deployment with 5 desired replicas is now been scaled up to 10 desired replicas, while the corresponding subscription is as same as before.

**Review by Garrybest — COMMENTED** [review]
Review is still going on.

File: pkg/controllers/apps/feedinventory/deployment/deployment.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+	// all the fields that we needed are backward compatible with apps/v1beta1, apps/v1beta2, extensions/v1beta1
+	deploy := &k8sappsv1.Deployment{}
+
+	if err := json.Unmarshal(rawData, deploy); err != nil {
+		return nil, appsapi.ReplicaRequirements{}, err
+	}
+
+	resourceLimits := map[corev1.ResourceName]resource.Quantity{}
+	resourceRequests := map[corev1.ResourceName]resource.Quantity{}
+	for _, container := range deploy.Spec.Template.Spec.Containers {
```
Comment (Garrybest) [inline pkg/controllers/apps/feedinventory/deployment/deployment.go:50]: Init containers should be considered. Please refer to https://kubernetes.io/docs/concepts/workloads/pods/init-containers/#resources.

File: pkg/controllers/apps/feedinventory/replicaset/replicaset.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+	// all the fields that we needed are backward compatible with extensions/v1beta1
+	rs := &k8sappsv1.ReplicaSet{}
+
+	if err := json.Unmarshal(rawData, rs); err != nil {
+		return nil, appsapi.ReplicaRequirements{}, err
+	}
+
+	resourceLimits := map[corev1.ResourceName]resource.Quantity{}
+	resourceRequests := map[corev1.ResourceName]resource.Quantity{}
+	for _, container := range rs.Spec.Template.Spec.Containers {
```
Comment (Garrybest) [inline pkg/controllers/apps/feedinventory/replicaset/replicaset.go:50]: Same as `deployment`.

File: pkg/controllers/apps/feedinventory/statefulset/statefulset.go [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+	// use apps/v1 for deserialization
+	// all the fields that we needed are backward compatible with apps/v1beta1, apps/v1beta2, extensions/v1beta1
+	sts := &k8sappsv1.StatefulSet{}
+
+	if err := json.Unmarshal(rawData, sts); err != nil {
+		return nil, appsapi.ReplicaRequirements{}, err
+	}
+
+	resourceLimits := map[corev1.ResourceName]resource.Quantity{}
+	resourceRequests := map[corev1.ResourceName]resource.Quantity{}
```
Comment (Garrybest) [inline pkg/controllers/apps/feedinventory/statefulset/statefulset.go:49]: Same as `deployment`.

File: pkg/controllers/apps/feedinventory/feedinventory.go [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+			Name:      sub.Name,
+			Namespace: sub.Namespace,
+			Labels: map[string]string{
+				known.ObjectCreatedByLabel: known.ClusternetHubName,
+				// add subscription info
+				known.ConfigSubscriptionNameLabel:      sub.Name,
+				known.ConfigSubscriptionNamespaceLabel: sub.Namespace,
+				known.ConfigSubscriptionUIDLabel:       string(sub.UID),
+			},
+			OwnerReferences: []metav1.OwnerReference{
```
Comment (Garrybest) [inline pkg/controllers/apps/feedinventory/feedinventory.go:269]: Maybe better to use `*metav1.NewControllerRef(sub, sub.GroupVersionKind())`

**Comment by Garrybest** [comment]
cc @qianjun1993

File: pkg/controllers/apps/feedinventory/feedinventory.go (hunk 1 above)
Comment (dixudx) [reply pkg/controllers/apps/feedinventory/feedinventory.go:302]: Yes.

File: pkg/controllers/apps/feedinventory/feedinventory.go (hunk 2 above)
Comment (dixudx) [reply pkg/controllers/apps/feedinventory/feedinventory.go]: Make sense. Addressed.

**Pull request merged**